File: apps/mac/Sources/DieterClient/SharedKV.swift
import CryptoKit
import DieterAPI
import Foundation
import GRPCCore
import SwiftProtobuf

package protocol SharedKVConnection: Sendable {
    func listKV(_ request: Dieter_V1_KVListRequest) async throws -> Dieter_V1_KVPage
    func getKV(_ request: Dieter_V1_KVRef) async throws -> Dieter_V1_KVEntry
    func putKV(_ request: Dieter_V1_KVPutRequest) async throws -> Dieter_V1_KVEntry
    func deleteKV(_ request: Dieter_V1_KVDeleteRequest) async throws -> Dieter_V1_KVEntry
    func moveKV(_ request: Dieter_V1_KVMoveRequest) async throws -> Dieter_V1_KVEntry
    func watchKV(_ request: Dieter_V1_KVWatchRequest, receive: @escaping @Sendable (Dieter_V1_KVFrame) async -> Void)
        async throws
}

extension DieterRPC: SharedKVConnection {
    package func listKV(_ request: Dieter_V1_KVListRequest) async throws -> Dieter_V1_KVPage {
        try await service.listKV(request: .init(message: request), options: Self.boundedUnaryCallOptions())
    }
    package func getKV(_ request: Dieter_V1_KVRef) async throws -> Dieter_V1_KVEntry {
        try await service.getKV(request: .init(message: request), options: Self.boundedUnaryCallOptions())
    }
    package func putKV(_ request: Dieter_V1_KVPutRequest) async throws -> Dieter_V1_KVEntry {
        try await service.putKV(request: .init(message: request), options: Self.boundedUnaryCallOptions())
    }
    package func deleteKV(_ request: Dieter_V1_KVDeleteRequest) async throws -> Dieter_V1_KVEntry {
        try await service.deleteKV(request: .init(message: request), options: Self.boundedUnaryCallOptions())
    }
    package func moveKV(_ request: Dieter_V1_KVMoveRequest) async throws -> Dieter_V1_KVEntry {
        try await service.moveKV(request: .init(message: request), options: Self.boundedUnaryCallOptions())
    }
    package func watchKV(
        _ request: Dieter_V1_KVWatchRequest, receive: @escaping @Sendable (Dieter_V1_KVFrame) async -> Void
    ) async throws {
        try await service.watchKV(request: .init(message: request)) { response in
            for try await frame in response.messages { try Task.checkCancellation(); await receive(frame) }
        }
    }
}

package struct SharedPosition: Codable, Equatable, Sendable {
    package var parent: String
    package var rank: String
    package init(parent: String, rank: String) { self.parent = parent; self.rank = rank }
}

/// One account subscription and durable outbox. A prepared request is immutable
/// until success or a definite rejection; uncertain requests remain daemon-bound.
@MainActor package final class SharedKV {
    struct Intent: Codable {
        var id = UUID().uuidString
        var key: String
        var value: Data?
        var deleted = false
        var parent: String?
        var after = ""
        var before = ""
        var requiresExisting = false
        var prepared: Data?
        var daemonID: String?
    }
    struct Cache: Codable {
        var entries: [String: Data] = [:]
        var pending: [Intent] = []
    }
    package private(set) var account = ""
    package private(set) var pendingCount = 0
    package private(set) var error: String?
    package var changed: (@MainActor () -> Void)?
    private let defaults: UserDefaults
    private let root: URL?
    private let namespace: String
    private var cache = Cache()
    private var rpc: (any SharedKVConnection)?
    private var daemonID = ""
    private var task: Task<Void, Never>?
    private var delivery: Task<Void, Never>?
    private var generation = 0
    private var replacement: [String: Data] = [:]
    package init(defaults: UserDefaults, root: URL? = nil, namespace: String = "navigation") {
        self.defaults = defaults; self.root = root; self.namespace = namespace
        account = defaults.string(forKey: "DieterSharedKV.activeAccount") ?? ""
        daemonID = defaults.string(forKey: "DieterSharedKV.activeDaemon") ?? ""
        cache = readCache().flatMap { try? JSONDecoder().decode(Cache.self, from: $0) } ?? Cache()
        pendingCount = cache.pending.count
    }
    deinit { task?.cancel(); delivery?.cancel() }
    private var cacheKey: String {
        "DieterSharedKV." + account + "." + namespace + (account == "local" ? "." + daemonID : "")
    }
    private var cacheURL: URL? {
        let digest = SHA256.hash(data: Data(cacheKey.utf8)).map { String(format: "%02x", $0) }.joined()
        return root?.appending(path: digest + ".json")
    }
    private func readCache() -> Data? {
        if let cacheURL { return try? Data(contentsOf: cacheURL) }
        return defaults.data(forKey: cacheKey)
    }
    private var entries: [String: Dieter_V1_KVEntry] {
        cache.entries.compactMapValues { try? Dieter_V1_KVEntry(serializedBytes: $0) }
    }
    package var values: [String: Data] {
        var result = entries.filter { !$0.value.deleted }.mapValues(\.valueJson)
        for intent in cache.pending {
            if intent.deleted {
                result.removeValue(forKey: intent.key)
            } else if let parent = intent.parent {
                let positions = result.compactMapValues { try? JSONDecoder().decode(SharedPosition.self, from: $0) }
                let prefix = String(intent.key.prefix { $0 != "." }) + "."
                let left =
                    positions[intent.after]?.rank
                    ?? (intent.before.isEmpty
                        ? positions.filter {
                            $0.key != intent.key && $0.key.hasPrefix(prefix) && $0.value.parent == parent
                        }.values.map(\.rank).max() ?? "" : "")
                let right = positions[intent.before]?.rank ?? ""
                let rank =
                    Self.between(left, right) + intent.id.lowercased().replacingOccurrences(of: "-", with: "") + "h"
                result[intent.key] = try? JSONEncoder().encode(SharedPosition(parent: parent, rank: rank))
            } else if !intent.requiresExisting || entries[intent.key]?.deleted != true {
                result[intent.key] = intent.value
            }
        }
        return result
    }
    private static func between(_ left: String, _ right: String) -> String {
        let digits = Array("0123456789abcdefghijklmnopqrstuvwxyz")
        let a = Array(left); var b = Array(right); var prefix = ""
        for index in 0..<512 {
            let low = index < a.count ? digits.firstIndex(of: a[index]) ?? 0 : 0
            let high = index < b.count ? digits.firstIndex(of: b[index]) ?? 35 : 35
            if high - low > 1 { return prefix + String(digits[(low + high) / 2]) }
            prefix.append(digits[low]); if high > low { b = [] }
        }
        return left + "h"
    }
    package func put<T: Encodable>(_ key: String, _ value: T, requiresExisting: Bool = false) {
        guard let data = try? JSONEncoder().encode(value) else { return }
        guard data.count <= 32 << 10 else {
            error = "Shared values must be at most 32 KiB."; changed?(); return
        }
        enqueue(Intent(key: key, value: data, requiresExisting: requiresExisting))
    }
    package func delete(_ key: String) { enqueue(Intent(key: key, deleted: true)) }
    package func move(_ key: String, parent: String = "", after: String = "", before: String = "") {
        enqueue(Intent(key: key, parent: parent, after: after, before: before))
    }
    private func enqueue(_ intent: Intent) {
        guard !account.isEmpty else {
            error = "Connect to an account before organizing navigation."; changed?(); return
        }
        guard cache.pending.count < 1024 else {
            error = "Navigation has 1,024 pending edits. Reconnect before editing more."; changed?(); return
        }
        cache.pending.append(intent)
        guard save() else { cache.pending.removeLast(); pendingCount = cache.pending.count; changed?(); return }
        changed?(); flush()
    }
    @discardableResult private func save() -> Bool {
        pendingCount = cache.pending.count
        defaults.set(account, forKey: "DieterSharedKV.activeAccount")
        defaults.set(daemonID, forKey: "DieterSharedKV.activeDaemon")
        if !account.isEmpty, let data = try? JSONEncoder().encode(cache) {
            do {
                if let root, let cacheURL {
                    try FileManager.default.createDirectory(
                        at: root, withIntermediateDirectories: true, attributes: [.posixPermissions: 0o700])
                    try data.write(to: cacheURL, options: .atomic)
                    try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: cacheURL.path)
                } else {
                    defaults.set(data, forKey: cacheKey)
                }
            } catch {
                self.error = "Could not persist navigation: \(error.localizedDescription)"
                delivery?.cancel(); rpc = nil; changed?(); return false
            }
        }
        return true
    }
    package func clearAccount() {
        bind(nil); account = ""; cache = Cache(); pendingCount = 0; error = nil; save(); changed?()
    }
    package func bind(_ connection: (any SharedKVConnection)?) {
        generation += 1; let token = generation
        task?.cancel(); delivery?.cancel(); delivery = nil; rpc = connection
        guard let connection else { return }
        task = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled && token == self.generation {
                do {
                    var request = Dieter_V1_KVListRequest(); request.namespace = self.namespace
                    let info = try await connection.listKV(request)
                    guard token == self.generation else { return }
                    if self.account != info.account || (info.account == "local" && self.daemonID != info.daemonID) {
                        self.account = info.account; self.daemonID = info.daemonID
                        self.cache =
                            self.readCache().flatMap { try? JSONDecoder().decode(Cache.self, from: $0) } ?? Cache()
                        self.changed?()
                        // Invalidate in-flight deliveries before subscribing to
                        // another account, even when the transport object survived.
                        self.bind(connection)
                        return
                    }
                    self.daemonID = info.daemonID
                    var watch = Dieter_V1_KVWatchRequest(); watch.namespace = self.namespace;
                    watch.account = info.account
                    // Replace only at caughtUp, so paginated bootstrap never flashes partial layouts.
                    try await connection.watchKV(watch) { [weak self] frame in
                        await self?.receive(frame, generation: token)
                    }
                } catch {
                    guard token == self.generation && !Task.isCancelled else { return }
                    self.error = String(describing: error); self.changed?()
                }
                try? await Task.sleep(for: .seconds(2))
            }
        }
    }
    private func receive(_ frame: Dieter_V1_KVFrame, generation token: Int) {
        guard token == generation, frame.account == account else { return }
        if frame.reset { replacement = [:] }
        for entry in frame.entries { replacement[entry.key] = try? entry.serializedData() }
        guard frame.caughtUp else { return }
        for (key, raw) in replacement {
            let incoming = try? Dieter_V1_KVEntry(serializedBytes: raw)
            let old = cache.entries[key].flatMap { try? Dieter_V1_KVEntry(serializedBytes: $0) }
            // Keep read-your-writes while a different replica catches up.
            let covers =
                old?.versions.allSatisfy { prior in
                    incoming?.versions.contains { next in
                        prior.clock.allSatisfy { actor, count in (next.clock[actor] ?? 0) >= count }
                    } == true
                } ?? true
            if covers { cache.entries[key] = raw }
        }
        error = nil; guard save() else { return }; changed?(); flush()
    }
    private static func covers(_ incoming: Dieter_V1_KVEntry?, _ prior: Dieter_V1_KVEntry) -> Bool {
        prior.versions.allSatisfy { old in
            incoming?.versions.contains { next in
                old.clock.allSatisfy { actor, count in (next.clock[actor] ?? 0) >= count }
            } == true
        }
    }
    private static func anchors(
        _ rpc: any SharedKVConnection, ref: Dieter_V1_KVRef, parent: String, after: String, before: String
    ) async throws -> (String, String) {
        func rank(_ key: String) async throws -> String? {
            guard !key.isEmpty else { return nil }
            var request = ref; request.key = key
            do {
                let entry = try await rpc.getKV(request)
                guard !entry.deleted,
                    let position = try? JSONDecoder().decode(SharedPosition.self, from: entry.valueJson),
                    position.parent == parent
                else { return nil }
                return position.rank
            } catch let e as RPCError where e.code == .notFound { return nil }
        }
        let left = try await rank(after), right = try await rank(before)
        if let left, let right, left >= right { return ("", before) }
        return (left == nil ? "" : after, right == nil ? "" : before)
    }
    private func flush() {
        guard delivery == nil, let rpc, !cache.pending.isEmpty else { return }
        let token = generation
        delivery = Task { [weak self] in
            guard let self else { return }
            defer { if token == self.generation { self.delivery = nil } }
            while !Task.isCancelled && token == self.generation && !self.cache.pending.isEmpty {
                do {
                    var intent = self.cache.pending[0]
                    if let accepting = intent.daemonID, accepting != self.daemonID {
                        self.error = "A pending navigation edit awaits its accepting machine."; self.changed?(); return
                    }
                    if intent.prepared == nil {
                        var ref = Dieter_V1_KVRef(); ref.namespace = self.namespace; ref.key = intent.key;
                        ref.account = self.account
                        var current: Dieter_V1_KVEntry?
                        do { current = try await rpc.getKV(ref) } catch let e as RPCError where e.code == .notFound {}
                        guard token == self.generation else { return }
                        if let prior = self.cache.entries[intent.key].flatMap({
                            try? Dieter_V1_KVEntry(serializedBytes: $0)
                        }), !Self.covers(current, prior) {
                            self.error = "Waiting for this machine to receive earlier navigation edits.";
                            self.changed?()
                            try await Task.sleep(for: .seconds(1)); continue
                        }
                        if intent.requiresExisting && (current == nil || current?.deleted == true) {
                            self.cache.pending.removeFirst(); guard self.save() else { return }; self.changed?();
                            continue
                        }
                        if let parent = intent.parent {
                            var request = Dieter_V1_KVMoveRequest(); request.ref = ref; request.parent = parent
                            let anchors = try await Self.anchors(
                                rpc, ref: ref, parent: parent, after: intent.after, before: intent.before)
                            guard token == self.generation else { return }
                            request.afterKey = anchors.0; request.beforeKey = anchors.1
                            request.expectedRevision = current?.revision ?? ""; request.operationID = intent.id;
                            request.daemonID = self.daemonID
                            intent.prepared = try request.serializedData()
                        } else if intent.deleted {
                            var request = Dieter_V1_KVDeleteRequest(); request.ref = ref
                            request.expectedRevision = current?.revision ?? ""; request.operationID = intent.id;
                            request.daemonID = self.daemonID
                            intent.prepared = try request.serializedData()
                        } else {
                            var request = Dieter_V1_KVPutRequest(); request.ref = ref;
                            request.valueJson = intent.value ?? Data()
                            request.expectedRevision = current?.revision ?? ""; request.operationID = intent.id;
                            request.daemonID = self.daemonID
                            intent.prepared = try request.serializedData()
                        }
                        intent.daemonID = self.daemonID; self.cache.pending[0] = intent;
                        guard self.save() else { return }
                    }
                    let result: Dieter_V1_KVEntry
                    if intent.parent != nil {
                        result = try await rpc.moveKV(.init(serializedBytes: intent.prepared!))
                    } else if intent.deleted {
                        result = try await rpc.deleteKV(.init(serializedBytes: intent.prepared!))
                    } else {
                        result = try await rpc.putKV(.init(serializedBytes: intent.prepared!))
                    }
                    guard token == self.generation else { return }
                    let prior = self.cache.entries[result.key].flatMap { try? Dieter_V1_KVEntry(serializedBytes: $0) }
                    if prior == nil || Self.covers(result, prior!) {
                        self.cache.entries[result.key] = try result.serializedData()
                    }
                    self.cache.pending.removeFirst(); self.error = nil; guard self.save() else { return };
                    self.changed?()
                } catch let e as RPCError where e.code == .aborted {
                    guard token == self.generation else { return }
                    self.cache.pending[0].prepared = nil; self.cache.pending[0].daemonID = nil
                    self.cache.pending[0].id = UUID().uuidString; guard self.save() else { return }
                    try? await Task.sleep(for: .milliseconds(250))
                } catch {
                    guard token == self.generation else { return }
                    self.error = String(describing: error); self.changed?()
                    try? await Task.sleep(for: .seconds(2))
                }
            }
        }
    }
}

File: apps/android/app/src/main/java/com/dbpprt/dieter/settings/SharedKV.kt
package com.dbpprt.dieter.settings

import android.content.SharedPreferences
import android.util.Base64
import com.dbpprt.dieter.data.DieterRepository
import com.dbpprt.dieter.v1.*
import com.google.protobuf.ByteString
import io.grpc.Status
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

/** Account cache and durable, daemon-bound outbox for portable navigation. */
class SharedKV(private val preferences: SharedPreferences, private val namespace: String = "navigation") {
    // State, protobuf/JSON projection and durable commits share one IO lane.
    // Network suspensions may interleave; each non-suspending mutation stays
    // atomic, as it did on Main, without blocking UI input or rendering.
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO.limitedParallelism(1))
    private val commands = Channel<() -> Unit>(1024)
    private var repository: DieterRepository? = null
    private var watch: Job? = null
    private var delivery: Job? = null
    private var generation = 0
    private var account = ""
    private var daemon = ""
    private var entries = mutableMapOf<String, KVEntry>()
    private var pending = mutableListOf<JSONObject>()
    private val _values = MutableStateFlow<Map<String, String>>(emptyMap())
    val values = _values.asStateFlow()
    private val _status = MutableStateFlow(SharedKVStatus())
    val status = _status.asStateFlow()
    private fun encode(bytes: ByteArray) = Base64.encodeToString(bytes, Base64.NO_WRAP)
    private fun decode(value: String) = Base64.decode(value, Base64.NO_WRAP)
    private val cacheKey get() = "account.$account.$namespace" + if (account == "local") ".$daemon" else ""
    private fun restore() {
        val saved = runCatching { JSONObject(preferences.getString(cacheKey, "{}")!!) }.getOrDefault(JSONObject())
        val records = saved.optJSONObject("entries") ?: JSONObject()
        entries = records.keys().asSequence().mapNotNull { key ->
            runCatching { key to KVEntry.parseFrom(decode(records.getString(key))) }.getOrNull()
        }.toMap().toMutableMap()
        val queue = saved.optJSONArray("pending") ?: JSONArray()
        pending = (0 until queue.length()).map { queue.getJSONObject(it) }.toMutableList()
        publish()
    }
    init {
        scope.launch {
            account = preferences.getString("activeAccount", "").orEmpty()
            daemon = preferences.getString("activeDaemon", "").orEmpty()
            if (account.isNotEmpty()) restore()
            for (command in commands) {
                try { command() }
                catch (error: Exception) { _status.value = _status.value.copy(error = error.message) }
            }
        }
    }

    private fun submit(command: () -> Unit) {
        if (commands.trySend(command).isFailure) {
            _status.value = _status.value.copy(error = "Navigation is busy. Try again after pending edits finish.")
        }
    }

    /** Completes after earlier commands and their durable writes, without waiting for network delivery. */
    internal suspend fun awaitPendingWrites() {
        val completed = CompletableDeferred<Unit>()
        commands.send { completed.complete(Unit) }
        completed.await()
    }

    suspend fun close() { commands.cancel(); scope.coroutineContext.job.cancelAndJoin() }

    // Read-modify-write projections must read after earlier queued edits.
    // Calculating their diff on Main would lose rapid successive reorderings.
    fun edit(transform: Editor.(Map<String, String>) -> Unit) = submit { Editor().transform(_values.value) }
    inner class Editor internal constructor() {
        fun put(key: String, value: Any, requiresExisting: Boolean = false) = putNow(key, value, requiresExisting)
        fun delete(key: String) = enqueue(JSONObject().put("key", key).put("deleted", true))
        fun move(key: String, parent: String = "", after: String = "", before: String = "") = enqueue(JSONObject()
            .put("key", key).put("parent", parent).put("after", after).put("before", before))
    }
    fun bind(connection: DieterRepository?) = submit { bindNow(connection) }
    private fun bindNow(connection: DieterRepository?) {
        generation++; val token = generation
        watch?.cancel(); delivery?.cancel(); delivery = null; repository = connection
        if (connection == null) return
        watch = scope.launch(start = CoroutineStart.LAZY) {
            while (isActive && token == generation) {
                try {
                    val info = connection.listKV(KVListRequest.newBuilder().setNamespace(namespace).build())
                    if (token != generation) return@launch
                    if (account != info.account || (account == "local" && daemon != info.daemonId)) {
                        account = info.account; daemon = info.daemonId; restore()
                        // A transport can survive an enrollment/account change.
                        // Invalidate old admissions before consuming another cache.
                        bindNow(connection); return@launch
                    }
                    daemon = info.daemonId
                    val replacement = mutableMapOf<String, KVEntry>()
                    connection.watchKV(KVWatchRequest.newBuilder().setNamespace(namespace).setAccount(account).build()).collect { frame ->
                        if (token != generation || frame.account != account) return@collect
                        if (frame.reset) replacement.clear()
                        frame.entriesList.forEach { replacement[it.key] = it }
                        if (frame.caughtUp) {
                            replacement.forEach { (key, incoming) ->
                                val covers = entries[key]?.versionsList?.all { prior ->
                                    incoming.versionsList.any { next -> prior.clockMap.all { (actor, count) -> (next.clockMap[actor] ?: 0) >= count } }
                                } ?: true
                                if (covers) entries[key] = incoming
                            }
                            if (save()) { publish(); flush() }
                        }
                    }
                } catch (e: CancellationException) { throw e }
                catch (e: Exception) { if (token == generation) _status.value = SharedKVStatus(pending.size, e.message) }
                delay(2_000)
            }
        }
        watch?.start()
    }
    fun clearAccount() = submit {
        bindNow(null)
        if (account.isNotEmpty()) {
            account = ""; daemon = ""; entries.clear(); pending.clear()
            preferences.edit().remove("activeAccount").remove("activeDaemon").commit()
        }
        publish()
    }
    fun put(key: String, value: Any, requiresExisting: Boolean = false) = submit { putNow(key, value, requiresExisting) }
    private fun putNow(key: String, value: Any, requiresExisting: Boolean) {
        val encoded = JSONArray().put(value).toString().let { it.substring(1, it.length - 1) }
        if (encoded.toByteArray(Charsets.UTF_8).size > 32 * 1024) {
            _status.value = SharedKVStatus(pending.size, "Shared values must be at most 32 KiB."); return
        }
        enqueue(JSONObject().put("key", key).put("value", encoded).put("requiresExisting", requiresExisting))
    }
    fun delete(key: String) = submit { enqueue(JSONObject().put("key", key).put("deleted", true)) }
    fun move(key: String, parent: String = "", after: String = "", before: String = "") = submit { enqueue(JSONObject()
        .put("key", key).put("parent", parent).put("after", after).put("before", before)) }
    private fun enqueue(intent: JSONObject) {
        if (account.isEmpty()) { _status.value = SharedKVStatus(pending.size, "Connect to an account before organizing navigation."); return }
        if (pending.size >= 1024) { _status.value = SharedKVStatus(pending.size, "Reconnect to deliver pending navigation edits."); return }
        intent.put("id", UUID.randomUUID().toString()); pending.add(intent)
        if (!save()) { pending.removeAt(pending.lastIndex); return }
        publish(); flush()
    }
    private fun save(): Boolean {
        if (account.isEmpty()) return true
        return try {
            val records = JSONObject(); entries.forEach { (key, entry) -> records.put(key, encode(entry.toByteArray())) }
            check(preferences.edit().putString("activeAccount", account).putString("activeDaemon", daemon).putString(cacheKey,
                JSONObject().put("entries", records).put("pending", JSONArray(pending)).toString()).commit()) { "Could not persist shared navigation" }
            true
        } catch (e: Exception) {
            _status.value = SharedKVStatus(pending.size, e.message); delivery?.cancel(); repository = null; false
        }
    }
    private fun publish() {
        val values = entries.filterValues { !it.deleted }.mapValues { it.value.valueJson.toStringUtf8() }.toMutableMap()
        pending.forEach { intent ->
            val key = intent.getString("key")
            when {
                intent.optBoolean("deleted") -> values.remove(key)
                intent.has("parent") -> {
                    val parent = intent.getString("parent")
                    val positions = values.mapNotNull { (k,v) -> runCatching { k to JSONObject(v) }.getOrNull() }.toMap()
                    val after = intent.optString("after"); val before = intent.optString("before")
                    val left = positions[after]?.optString("rank") ?: if (before.isEmpty()) positions.filter {
                        it.key != key && it.key.substringBefore('.') == key.substringBefore('.') && it.value.optString("parent") == parent
                    }.values.map { it.optString("rank") }.maxOrNull().orEmpty() else ""
                    val right = positions[before]?.optString("rank").orEmpty()
                    values[key] = JSONObject().put("parent", parent).put("rank", between(left,right) + intent.getString("id").replace("-", "") + "h").toString()
                }
                !intent.optBoolean("requiresExisting") || entries[key]?.deleted != true -> values[key] = intent.getString("value")
            }
        }
        _values.value = values; _status.value = SharedKVStatus(pending.size)
    }
    private fun flush() {
        val connection = repository ?: return
        if (delivery?.isActive == true || pending.isEmpty()) return
        val token = generation
        delivery = scope.launch {
            while (isActive && token == generation && pending.isNotEmpty()) {
                try {
                    val intent = pending.first()
                    if (intent.has("daemon") && intent.getString("daemon") != daemon) {
                        _status.value = SharedKVStatus(pending.size, "A pending navigation edit awaits its accepting machine."); return@launch
                    }
                    if (!intent.has("prepared")) {
                        val ref = KVRef.newBuilder().setNamespace(namespace).setKey(intent.getString("key")).setAccount(account).build()
                        val current = try { connection.getKV(ref) } catch(e: Exception) {
                            if (Status.fromThrowable(e).code == Status.Code.NOT_FOUND) null else throw e
                        }
                        if (token != generation) return@launch
                        if (entries[ref.key]?.let { !covers(current, it) } == true) {
                            _status.value = SharedKVStatus(pending.size, "Waiting for this machine to receive earlier navigation edits.")
                            delay(1_000); continue
                        }
                        if (intent.optBoolean("requiresExisting") && (current == null || current.deleted)) {
                            pending.removeAt(0); if (!save()) return@launch; publish(); continue
                        }
                        val revision = current?.revision.orEmpty(); val id = intent.getString("id")
                        var after = intent.optString("after"); var before = intent.optString("before")
                        if (intent.has("parent")) {
                            suspend fun rank(key: String): String? {
                                if (key.isEmpty()) return null
                                val entry = try { connection.getKV(ref.toBuilder().setKey(key).build()) }
                                    catch (e: Exception) { if (Status.fromThrowable(e).code == Status.Code.NOT_FOUND) return null else throw e }
                                val position = runCatching { JSONObject(entry.valueJson.toStringUtf8()) }.getOrNull()
                                return position?.takeIf { !entry.deleted && it.optString("parent") == intent.getString("parent") }?.optString("rank")
                            }
                            val left = rank(after); val right = rank(before)
                            if (left == null || (right != null && left >= right)) after = ""
                            if (right == null) before = ""
                        }
                        if (token != generation) return@launch
                        val bytes = when {
                            intent.has("parent") -> KVMoveRequest.newBuilder().setRef(ref).setParent(intent.getString("parent"))
                                .setAfterKey(after).setBeforeKey(before)
                                .setExpectedRevision(revision).setOperationId(id).setDaemonId(daemon).build().toByteArray()
                            intent.optBoolean("deleted") -> KVDeleteRequest.newBuilder().setRef(ref).setExpectedRevision(revision)
                                .setOperationId(id).setDaemonId(daemon).build().toByteArray()
                            else -> KVPutRequest.newBuilder().setRef(ref).setValueJson(ByteString.copyFromUtf8(intent.getString("value")))
                                .setExpectedRevision(revision).setOperationId(id).setDaemonId(daemon).build().toByteArray()
                        }
                        intent.put("prepared", encode(bytes)).put("daemon", daemon); if (!save()) return@launch
                    }
                    val bytes = decode(intent.getString("prepared"))
                    val result = when {
                        intent.has("parent") -> connection.moveKV(KVMoveRequest.parseFrom(bytes))
                        intent.optBoolean("deleted") -> connection.deleteKV(KVDeleteRequest.parseFrom(bytes))
                        else -> connection.putKV(KVPutRequest.parseFrom(bytes))
                    }
                    if (token != generation) return@launch
                    if (entries[result.key]?.let { covers(result,it) } != false) entries[result.key] = result
                    pending.removeAt(0); if (!save()) return@launch; publish()
                } catch (e: CancellationException) { throw e }
                catch (e: Exception) {
                    if (token != generation) return@launch
                    if (Status.fromThrowable(e).code == Status.Code.ABORTED) {
                        pending.first().apply { remove("prepared"); remove("daemon"); put("id", UUID.randomUUID().toString()) }; if (!save()) return@launch
                    } else _status.value = SharedKVStatus(pending.size, e.message)
                    delay(2_000)
                }
            }
        }
    }
    private fun covers(incoming: KVEntry?, prior: KVEntry): Boolean = prior.versionsList.all { old ->
        incoming?.versionsList?.any { next -> old.clockMap.all { (actor,count) -> (next.clockMap[actor] ?: 0) >= count } } == true
    }
    companion object {
        private fun between(left: String, right: String): String {
            val alphabet = "0123456789abcdefghijklmnopqrstuvwxyz"; val prefix = StringBuilder(); var upper = right
            repeat(512) { index ->
                val lo = left.getOrNull(index)?.let { alphabet.indexOf(it).coerceAtLeast(0) } ?: 0
                val hi = upper.getOrNull(index)?.let { alphabet.indexOf(it).coerceAtLeast(0) } ?: 35
                if (hi-lo > 1) return prefix.append(alphabet[(lo+hi)/2]).toString()
                prefix.append(alphabet[lo]); if (hi > lo) upper = ""
            }
            return left + "h"
        }
    }
}
data class SharedKVStatus(val pending: Int = 0, val error: String? = null)
